pkg/debugger/debugger.go

```diff
@@ -0,0 +1,202 @@
+package debugger
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/filecoin-project/mir/pkg/eventlog"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/logging"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+const (
+	ReadBufferSize  = 1024
+	WriteBufferSize = 1024
+)
+
+type WSWriter struct {
+	// ... websocket server variables ...
+	conn        *websocket.Conn
+	upgrader    websocket.Upgrader
+	eventSignal chan map[string]interface{}
+	WSMessage   struct {
+		Type  string `json:"Type"`
+		Value string `json:"Value"`
+	}
+}
+
+// InterceptorInit initializes the interceptor according to input boolean
+// The interceptor is set to nil if the user decides to not use the debugger
+func InterceptorInit(
+	ownID t.NodeID,
+	port string,
+) (*eventlog.Recorder, error) {
+	// writerFactory creates and returns a WebSocket-based event writer
+	writerFactory := func(_ string, ownID t.NodeID, _ logging.Logger) (eventlog.EventWriter, error) {
+		return newWSWriter(fmt.Sprintf(":%s", port)), nil
+	}
+
+	var interceptor *eventlog.Recorder
+	var err error
+	interceptor, err = eventlog.NewRecorder(
+		ownID,
+		fmt.Sprintf("./node%s", ownID),
+		logging.ConsoleInfoLogger,
+		eventlog.EventWriterOpt(writerFactory),
+		eventlog.SyncWriteOpt(),
+	)
+	if err != nil {
+		panic(err)
+	}
+	return interceptor, err
+}
+
+// Flush does nothing at the moment
+func (wsw *WSWriter) Flush() error {
+	return nil
+}
+
+// Close closes the connection
+func (wsw *WSWriter) Close() error {
+	if wsw.conn == nil {
+		return nil
+	}
+	return wsw.conn.Close()
+}
+
+// Write sends every event to the frontend and then wits for a message detailing how to proceed with that event
+// The returned EventList contains the accepted events
+func (wsw *WSWriter) Write(list *events.EventList, _ int64) (*events.EventList, error) {
+	for wsw.conn == nil {
+		fmt.Println("No connection.")
```

Comment:
Use a logger instead here. A pointer to a logger can be saved in a field of the `WSWriter`, set at creation, and accessed here through `wsw.logger`, for example.